packages/typir/src/features/printing.ts

```diff
@@ -60,41 +62,47 @@ export class DefaultTypeConflictPrinter implements ProblemPrinter {
         const left = problem.expected;
         const right = problem.actual;
         let result = '';
-        if (typeof problem.index === 'number') {
-            result = result + `At index ${problem.index}, `;
+        if (problem.propertyName) {
+            if (problem.propertyIndex) {
+                result = result + `For property '${problem.propertyName} at index ${problem.propertyIndex}', `;
+            } else {
+                result = result + `For property '${problem.propertyName}', `;
+            }
+        } else if (problem.propertyIndex) {
+            result = result + `At index ${problem.propertyIndex}, `;
         } else {
-            result = result + `For property '${problem.index}', `;
+            result = result + 'At an unknown location, ';
         }
         if (left !== undefined && right !== undefined) {
-            result = result + `the types '${this.printType(left)}' and '${this.printType(right)}' do not match.`;
+            result = result + `the types '${this.printTypeName(left)}' and '${this.printTypeName(right)}' do not match.`;
         } else if (left !== undefined && right === undefined) {
-            result = result + `the type '${this.printType(left)}' on the left has no opposite type on the right to match with.`;
+            result = result + `the type '${this.printTypeName(left)}' on the left has no opposite type on the right to match with.`;
         } else if (left === undefined && right !== undefined) {
-            result = result + `there is no type on the left to match with the type '${this.printType(right)}' on the right.`;
+            result = result + `there is no type on the left to match with the type '${this.printTypeName(right)}' on the right.`;
         } else {
-            throw new Error();
+            result = result + 'both types are unclear.';
         }
         result = this.printIndentation(result, level);
```

Comment:
As discussed, I simplified the code a bit, but decided against extracting/reusing the last two lines in the methods, since ...
- there is one method with only a single line (exception)
- it makes the customization/overriding of single methods a bit more complicated